esmf_regrid/esmf_regridder.py

```diff
@@ -290,27 +290,25 @@ def regrid(self, src_array, mdtol=1):
             A numpy array whose shape is compatible with self.tgt.
 
         """
-        # TODO implement masked array handling similar to other iris regridders.
-
-        # A rudimentary filter is applied to mask data which is mapped from an
-        # insufficiently large source. This currently only accounts for discrepancies
-        # between the source and target grid/mesh geometries and does not account for
-        # masked data, though it ought to be possible to extend the functionality to
-        # handle masked data.
-        #
-        # Note that ESMPy is also able to handle masked data. It is worth investigating
-        # how this affects the mathematics and if it can be replicated after the fact
-        # using just the weights or if ESMF is doing something we want access to.
-        weight_sums = np.array(self.weight_matrix.sum(axis=1)).flatten()
+        src_inverted_mask = np.where(ma.getmaskarray(src_array), 0, 1)
+        src_inverted_mask = self.src._flatten_array(src_inverted_mask)
+        weight_sums = self.weight_matrix * src_inverted_mask
         # Set the minimum mdtol to be slightly higher than 0 to account for rounding
         # errors.
         mdtol = max(mdtol, 1e-8)
-        tgt_mask = weight_sums >= 1 - mdtol
+        tgt_mask = weight_sums > 1 - mdtol
         masked_weight_sums = weight_sums * tgt_mask.astype(int)
```

Comment:
Is it too late to ask why we cast this to `int`?

Reply:
I think the need for casting this was due to tgt_mask being bool type which didn't multiply well. Casting it as int may well be slightly inappropriate, my intuition was that it would cause less ambiguity with levels of precision, but I'm not really sure what the consequences of casting it to int vs float would be.

Reply:
I think the multiplication will cast the mask to the type of `weight_sums` right away. In principle, multiplication with a bool array should be fine.